README.md
ssh-keydgen [![Go Report Card](https://goreportcard.com/badge/github.com/cornfeedhobo/ssh-keydgen)](https://goreportcard.com/report/github.com/cornfeedhobo/ssh-keydgen) [![Build Status](https://travis-ci.org/cornfeedhobo/ssh-keydgen.svg?branch=master)](https://travis-ci.org/cornfeedhobo/ssh-keydgen) [![Github All Releases](https://img.shields.io/github/downloads/cornfeedhobo/ssh-keydgen/total.svg)](https://github.com/cornfeedhobo/ssh-keydgen/releases)
===========

Generate _Deterministic_ SSH keys

```text
NAME:
   ssh-keydgen - deterministic authentication key generation

USAGE:
   ssh-keydgen [[-t <type>] [-b <bits>] [-c <curve>] [-f <filename>] [-a <rounds>] [--at <time>] [--am <memory>] [--as <seedphrase>] [--aa]]

AUTHOR:
   cornfeedhobo

GLOBAL OPTIONS:
   -t type          Specifies the type of key to create. The possible values are "dsa", "ecdsa", "rsa", or "ed25519". (default: "rsa")
   -b bits          Specifies the number of bits in the key to create. Possible values are restricted by key type. (default: 2048)
   -c curve         Specifies the elliptic curve to use. The possible values are 256, 384, or 521. (default: 256)
   -f filename      Specifies the filename of the key file.
   -a rounds        Specifies the number of hashing rounds applied during key generation. (default: 1000)
   --at time        Specifies the time parameter for the Argon2 function. (default: 3)
   --am memory      Specifies the memory parameter for the Argon2 function. (default: 16384)
   --ap threads     Specifies the threads or parallelism for the Argon2 function. (default: 1)
   --as seedphrase  Provides the deterministic seedphrase.
   --aa             Add the generated key to the running ssh-agent.

COPYRIGHT:
   (c) 2018 cornfeedhobo
```



## Usage

1) Generate your keys

   ```bash
   keydgen -f path/to/deterministic_key
   ls -lh path/to/deterministic_key*
   ```

2) Allow time to pass, hoping an emergency does not arise when you have no access to your keys ...

   _If_ the time comes where you need access but can't get to your keys, you can then obtain this
   utility and re-generate, or even directly add your key to a running `ssh-agent`.

   ```bash
   ssh-keydgen --aa
   ```

3) Profit!



## FAQ

### What Go versions are supported?

Go 1.9 or later


### How can I verify the generated key is valid?

Until there are more implementations of this generation scheme, you can
at least verify the private key is usable and the public key matches what
openssh generates.

```bash
cat path/to/deterministic_key.pub
ssh-keygen -y -f path/to/deterministic_key
```

If the above outputs don't match, the public key was not generated properly.
If you are prompted for a password, the private key was not generated properly.


### How can I encrypt my key after generation?

```bash
ssh-keygen -p -f path/to/deterministic_key
```



## Is it any good?

[Yes](http://news.ycombinator.com/item?id=3067434)


CHANGELOG
* Sat Sept 8 2018 cornfeedhobo 0.4.1
- Add seedphrase verification
- Update dependencies

* Thu Jun 21 2018 cornfeedhobo 0.4.0
- Switch to argon2 and use pbkdf2 for seed and salt derivation
- Refactor and add documentation

* Mon Jun 4 2018 cornfeedhobo 0.3.0
- Add ssh-keygen test to validate output

* Mon Nov 20 2017 cornfeedhobo 0.2.0
- Refactor key management to a struct-based approach
- Drop printing to stdout and switch to writing files like OpenSSH's ssh-keygen
- Refactor resource names
- Style clean up
- Add configurable work factor
- Add tests

* Sun Nov 12 2017 cornfeedhobo 0.1.0
- Initial draft


main_test.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"

	"github.com/cornfeedhobo/ssh-keydgen/keygen"
	"github.com/cornfeedhobo/ssh-keydgen/slowseeder"
)

func TestKeydgen(t *testing.T) {

	cases := []*keygen.Keydgen{
		{
			Type: keygen.ED25519,
		},
		{
			Type:  keygen.ECDSA,
			Curve: 256,
		},
		{
			Type:  keygen.ECDSA,
			Curve: 384,
		},
		{
			Type:  keygen.ECDSA,
			Curve: 521,
		},
		{
			Type: keygen.RSA,
			Bits: 2048,
		},
		{
			Type: keygen.RSA,
			Bits: 4096,
		},
		{
			Type: keygen.DSA,
			Bits: 1024,
		},
		{
			Type: keygen.DSA,
			Bits: 2048,
		},
		{
			Type: keygen.DSA,
			Bits: 3072,
		},
	}

	for _, k := range cases {

		var name = k.Type
		switch k.Type {
		case keygen.RSA, keygen.DSA:
			name += fmt.Sprintf("_%d", k.Bits)
		case keygen.ECDSA:
			name += fmt.Sprintf("_%d", k.Curve)
		}

		t.Run(name, func(t *testing.T) {

			fmt.Print(name)
			start := time.Now()

			// use small parameters to keep tests short
			d, err := slowseeder.New([]byte("keygen"), 1, 1, 512, 1)
			if err != nil {
				t.Fatal(err)
			}

			_, err = k.GenerateKey(d)
			if err != nil {
				t.Fatal(err)
			}

			privBytes, err := k.MarshalPrivateKey()
			if err != nil {
				t.Fatal(err)
			}

			pubBytes, err := k.MarshalPublicKey()
			if err != nil {
				t.Fatal(err)
			}

			filename, err := filepath.Abs(name)
			if err != nil {
				t.Fatal(err)
			}

			err = writeKeyToFile(k, filename)
			if err != nil {
				t.Fatal(err)
			}

			cmd := exec.Command("ssh-keygen", "-y", "-f", filename)
			outPipe, _ := cmd.StdoutPipe()
			errPipe, _ := cmd.StderrPipe()
			err = cmd.Start()
			if err != nil {
				t.Fatal(err)
			}

			stdout, _ := ioutil.ReadAll(outPipe)
			stderr, _ := ioutil.ReadAll(errPipe)

			if !bytes.Equal(pubBytes, stdout) {
				msg := "Unable to verify generated public key with ssh-keygen"
				msg += "\n\nGenerated Private Key:\n" + string(privBytes)
				msg += "\n\nGenerated Public Key:\n" + string(pubBytes)
				msg += "\n\nStdout:\n" + string(stdout)
				msg += "\n\nStderr:\n" + string(stderr)
				t.Fatal(msg)
			}

			// don't defer in case inspection needs to be done with a failure
			os.Remove(filename)
			os.Remove(filename + ".pub")

			fmt.Printf(" PASS %s\n", time.Since(start))

		})
	}
}


keydgen_test.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"testing"
	"time"

	"github.com/google/goexpect"
)

func TestKeydgen_GenerateKey(t *testing.T) {

	WorkFactor = 1024 * 1

	cases := []struct {
		name          string
		expectFailure bool
		keydgen       *Keydgen
	}{
		// DSA
		{
			name: "DSA_InvalidKeyLength",
			keydgen: &Keydgen{
				Type: DSA,
				Bits: 100,
				Seed: []byte("keydgen"),
			},
			expectFailure: true,
		},
		{
			name: "DSA_1024",
			keydgen: &Keydgen{
				Type: DSA,
				Bits: 1024,
				Seed: []byte("keydgen"),
			},
		},
		{
			name: "DSA_2048",
			keydgen: &Keydgen{
				Type: DSA,
				Bits: 2048,
				Seed: []byte("keydgen"),
			},
		},
		{
			name: "DSA_3072",
			keydgen: &Keydgen{
				Type: DSA,
				Bits: 3072,
				Seed: []byte("keydgen"),
			},
		},
		// ECDSA
		{
			name: "ECDSA_InvalidCurve",
			keydgen: &Keydgen{
				Type:  ECDSA,
				Curve: 128,
				Seed:  []byte("keydgen"),
			},
			expectFailure: true,
		},
		{
			name: "ECDSA_256",
			keydgen: &Keydgen{
				Type:  ECDSA,
				Curve: 256,
				Seed:  []byte("keydgen"),
			},
		},
		{
			name: "ECDSA_384",
			keydgen: &Keydgen{
				Type:  ECDSA,
				Curve: 384,
				Seed:  []byte("keydgen"),
			},
		},
		{
			name: "ECDSA_521",
			keydgen: &Keydgen{
				Type:  ECDSA,
				Curve: 521,
				Seed:  []byte("keydgen"),
			},
		},
		// RSA
		{
			name: "RSA_2048",
			keydgen: &Keydgen{
				Type: RSA,
				Bits: 2048,
				Seed: []byte("keydgen"),
			},
		},
		// ED25519
		{
			name: "ED25519",
			keydgen: &Keydgen{
				Type: ED25519,
				Seed: []byte("keydgen"),
			},
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {

			start := time.Now()
			fmt.Print(c.name)

			if _, err := c.keydgen.GenerateKey(); err != nil && !c.expectFailure {
				t.Fatal(err)
			}

			if !c.expectFailure {

				pub, err := c.keydgen.MarshalPublicKey()
				if err != nil {
					t.Fatal(err)
				}

				filename := fmt.Sprintf("keydgen_test_%s_", c.keydgen.Type)
				if c.keydgen.Type == RSA || c.keydgen.Type == DSA {
					filename += fmt.Sprintf("%d_", c.keydgen.Bits)
				} else if c.keydgen.Type == ECDSA {
					filename += fmt.Sprintf("%d_", c.keydgen.Curve)
				}

				tmpFile, err := ioutil.TempFile("", filename)
				if err != nil {
					t.Fatal(err)
				}

				if err := writeKeyToFile(c.keydgen, tmpFile.Name()); err != nil {
					t.Fatal(err)
				}
				defer func() {
					os.Remove(tmpFile.Name())
					os.Remove(tmpFile.Name() + ".pub")
				}()

				e, _, err := expect.Spawn("ssh-keygen -y -f "+tmpFile.Name(), -1)
				if err != nil {
					t.Fatal(err)
				}
				defer e.Close()

				expr := regexp.MustCompilePOSIX(regexp.QuoteMeta(string(pub)))
				if out, _, err := e.Expect(expr, -1); err != nil {
					t.Fatal("unable to verify generated public key with ssh-keygen: " + out)
				}

			}

			fmt.Printf(" PASS %s\n", time.Since(start))

		})
	}

}


slowseeder/slowseeder_test.go
package slowseeder

import (
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"
)

func Example_generateRSA() {
	r, _ := New([]byte("slowseeder"), 1000, 3, 1024*16, 1)
	k, _ := rsa.GenerateKey(r, 2048)
	e := pem.EncodeToMemory(&pem.Block{
		Type:  "PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(k),
	})
	fmt.Println(string(e))
}


keygen/keygen.go
// Package keygen package is a single entry point for generating OpenSSH keys
package keygen

import (
	"bytes"
	"crypto/dsa"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rsa"
	"crypto/x509"
	"encoding/asn1"
	"encoding/pem"
	"errors"
	"io"
	"math/big"

	"github.com/mikesmitty/edkey"
	"golang.org/x/crypto/ed25519"
	"golang.org/x/crypto/ssh"
)

// These constants represent the support key types
const (
	DSA     = "dsa"
	ECDSA   = "ecdsa"
	RSA     = "rsa"
	ED25519 = "ed25519"
)

var (
	// ErrUnsupportedKeyType is the error returned when an unsupported key type is requested
	ErrUnsupportedKeyType = errors.New("unsupported key type")

	// ErrUnsupportedKeyLength is the error returned when an invalid key length is supplied for the key type
	ErrUnsupportedKeyLength = errors.New("invalid key length")

	// ErrUnsuppontedCurve is the error returned when generating an ECDSA key and an invalid curve is requested
	ErrUnsuppontedCurve = errors.New("only P-256, P-384 and P-521 EC keys are supported")
)

// Keydgen represents an OpenSSH key generator
type Keydgen struct {
	Type  string
	Bits  uint16
	Curve uint16

	privateKey interface{}
}

func (k *Keydgen) generateDSA(rand io.Reader) (interface{}, error) {

	var (
		size   dsa.ParameterSizes
		params = new(dsa.Parameters)
		key    = new(dsa.PrivateKey)
	)

	switch k.Bits {
	case 1024:
		size = dsa.L1024N160
	case 2048:
		size = dsa.L2048N256
	case 3072:
		size = dsa.L3072N256
	default:
		return nil, ErrUnsupportedKeyLength
	}

	if err := dsa.GenerateParameters(params, rand, size); err != nil {
		return nil, err
	}
	key.Parameters = *params

	err := dsa.GenerateKey(key, rand)

	return key, err

}

func (k *Keydgen) generateECDSA(rand io.Reader) (interface{}, error) {

	var c elliptic.Curve

	switch k.Curve {
	case 256:
		c = elliptic.P256()
	case 384:
		c = elliptic.P384()
	case 521:
		c = elliptic.P521()
	default:
		return nil, ErrUnsuppontedCurve
	}

	return ecdsa.GenerateKey(c, rand)

}

// GenerateKey generates and/or returns a private key
func (k *Keydgen) GenerateKey(rand io.Reader) (key interface{}, err error) {

	if k.privateKey != nil {
		return k.privateKey, nil
	}

	switch k.Type {
	case DSA:
		k.privateKey, err = k.generateDSA(rand)
	case ECDSA:
		k.privateKey, err = k.generateECDSA(rand)
	case RSA:
		k.privateKey, err = rsa.GenerateKey(rand, int(k.Bits))
	case ED25519:
		_, k.privateKey, err = ed25519.GenerateKey(rand)
	default:
		return nil, ErrUnsupportedKeyType
	}

	return k.privateKey, err

}

// MarshalPrivateKey returns an OpenSSH formatted private key
func (k *Keydgen) MarshalPrivateKey() ([]byte, error) {

	if k.privateKey == nil {
		panic("private key not hasn't been generated yet")
	}

	var (
		block *pem.Block
		buf   = bytes.NewBuffer(nil)
		err   error
	)

	switch k.Type {

	case DSA:
		block = &pem.Block{Type: "DSA PRIVATE KEY"}
		block.Bytes, err = asn1.Marshal(struct {
			Version       int
			P, Q, G, Y, X *big.Int
		}{
			P: k.privateKey.(*dsa.PrivateKey).P,
			Q: k.privateKey.(*dsa.PrivateKey).Q,
			G: k.privateKey.(*dsa.PrivateKey).G,
			Y: k.privateKey.(*dsa.PrivateKey).Y,
			X: k.privateKey.(*dsa.PrivateKey).X,
		})

	case ECDSA:
		block = &pem.Block{Type: "EC PRIVATE KEY"}
		block.Bytes, err = x509.MarshalECPrivateKey(k.privateKey.(*ecdsa.PrivateKey))

	case RSA:
		block = &pem.Block{
			Type:  "RSA PRIVATE KEY",
			Bytes: x509.MarshalPKCS1PrivateKey(k.privateKey.(*rsa.PrivateKey)),
		}

	case ED25519:
		block = &pem.Block{
			Type:  "OPENSSH PRIVATE KEY",
			Bytes: edkey.MarshalED25519PrivateKey(k.privateKey.(ed25519.PrivateKey)),
		}

	default:
		return nil, ErrUnsupportedKeyType

	}

	if err := pem.Encode(buf, block); err != nil {
		return nil, err
	}

	return buf.Bytes(), err

}

// MarshalPublicKey returns an OpenSSH formatted public key
func (k *Keydgen) MarshalPublicKey() ([]byte, error) {

	if k.privateKey == nil {
		panic("private key has not been generated yet")
	}

	var (
		pubKey ssh.PublicKey
		err    error
	)

	switch k.Type {

	case DSA:
		var pub = &k.privateKey.(*dsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case ECDSA:
		var pub = &k.privateKey.(*ecdsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case RSA:
		var pub = &k.privateKey.(*rsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case ED25519:
		var pub = k.privateKey.(ed25519.PrivateKey).Public().(ed25519.PublicKey)
		pubKey, err = ssh.NewPublicKey(pub)

	default:
		err = ErrUnsupportedKeyType

	}

	return ssh.MarshalAuthorizedKey(pubKey), err

}


keydgen/keydgen.go
package keydgen

import (
	"bytes"
	"crypto/dsa"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rsa"
	"crypto/x509"
	"encoding/asn1"
	"encoding/pem"
	"errors"
	"io"
	"math/big"

	"github.com/mikesmitty/edkey"
	"golang.org/x/crypto/ed25519"
	"golang.org/x/crypto/ssh"
)

// These constants represent the support key types
const (
	DSA     = "dsa"
	ECDSA   = "ecdsa"
	RSA     = "rsa"
	ED25519 = "ed25519"
)

var (
	// ErrUnsupportedKeyType is the error returned when an unsupported key type is requested
	ErrUnsupportedKeyType = errors.New("unsupported key type")

	// ErrUnsupportedKeyLength is the error returned when an invalid key length is supplied for the key type
	ErrUnsupportedKeyLength = errors.New("invalid key length")

	// ErrUnsuppontedCurve is the error returned when generating an ECDSA key and an invalid curve is requested
	ErrUnsuppontedCurve = errors.New("only P-256, P-384 and P-521 EC keys are supported")
)

// Keydgen represents a deterministic OpenSSH key generator
type Keydgen struct {
	Type  string
	Bits  uint16
	Curve uint16
	//Seed  []byte
	//Rounds  uint32
	//Time    uint32
	//Memory  uint32
	//Threads uint8

	privateKey interface{}
}

func (k *Keydgen) generateDSA(rand io.Reader) (interface{}, error) {

	var (
		size   dsa.ParameterSizes
		params = new(dsa.Parameters)
		key    = new(dsa.PrivateKey)
	)

	switch k.Bits {
	case 1024:
		size = dsa.L1024N160
	case 2048:
		size = dsa.L2048N256
	case 3072:
		size = dsa.L3072N256
	default:
		return nil, ErrUnsupportedKeyLength
	}

	if err := dsa.GenerateParameters(params, rand, size); err != nil {
		return nil, err
	}
	key.Parameters = *params

	err := dsa.GenerateKey(key, rand)

	return key, err

}

func (k *Keydgen) generateECDSA(rand io.Reader) (interface{}, error) {

	var c elliptic.Curve

	switch k.Curve {
	case 256:
		c = elliptic.P256()
	case 384:
		c = elliptic.P384()
	case 521:
		c = elliptic.P521()
	default:
		return nil, ErrUnsuppontedCurve
	}

	return ecdsa.GenerateKey(c, rand)

}

// GenerateKey generates and/or returns a private key
func (k *Keydgen) GenerateKey(rand io.Reader) (key interface{}, err error) {

	if k.privateKey != nil {
		return k.privateKey, nil
	}

	switch k.Type {
	case DSA:
		k.privateKey, err = k.generateDSA(rand)
	case ECDSA:
		k.privateKey, err = k.generateECDSA(rand)
	case RSA:
		k.privateKey, err = rsa.GenerateKey(rand, int(k.Bits))
	case ED25519:
		_, k.privateKey, err = ed25519.GenerateKey(rand)
	default:
		return nil, ErrUnsupportedKeyType
	}

	return k.privateKey, err

}

// MarshalPrivateKey returns an OpenSSH formatted private key
func (k *Keydgen) MarshalPrivateKey() ([]byte, error) {

	if k.privateKey == nil {
		panic("private key not hasn't been generated yet")
	}

	var (
		block *pem.Block
		buf   = bytes.NewBuffer(nil)
		err   error
	)

	switch k.Type {

	case DSA:
		block = &pem.Block{Type: "DSA PRIVATE KEY"}
		block.Bytes, err = asn1.Marshal(struct {
			Version       int
			P, Q, G, Y, X *big.Int
		}{
			P: k.privateKey.(*dsa.PrivateKey).P,
			Q: k.privateKey.(*dsa.PrivateKey).Q,
			G: k.privateKey.(*dsa.PrivateKey).G,
			Y: k.privateKey.(*dsa.PrivateKey).Y,
			X: k.privateKey.(*dsa.PrivateKey).X,
		})

	case ECDSA:
		block = &pem.Block{Type: "EC PRIVATE KEY"}
		block.Bytes, err = x509.MarshalECPrivateKey(k.privateKey.(*ecdsa.PrivateKey))

	case RSA:
		block = &pem.Block{
			Type:  "RSA PRIVATE KEY",
			Bytes: x509.MarshalPKCS1PrivateKey(k.privateKey.(*rsa.PrivateKey)),
		}

	case ED25519:
		block = &pem.Block{
			Type:  "OPENSSH PRIVATE KEY",
			Bytes: edkey.MarshalED25519PrivateKey(k.privateKey.(ed25519.PrivateKey)),
		}

	default:
		return nil, ErrUnsupportedKeyType

	}

	if err := pem.Encode(buf, block); err != nil {
		return nil, err
	}

	return buf.Bytes(), err

}

// MarshalPublicKey returns an OpenSSH formatted public key
func (k *Keydgen) MarshalPublicKey() ([]byte, error) {

	if k.privateKey == nil {
		panic("private key has not been generated yet")
	}

	var (
		pubKey ssh.PublicKey
		err    error
	)

	switch k.Type {

	case DSA:
		var pub = &k.privateKey.(*dsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case ECDSA:
		var pub = &k.privateKey.(*ecdsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case RSA:
		var pub = &k.privateKey.(*rsa.PrivateKey).PublicKey
		pubKey, err = ssh.NewPublicKey(pub)

	case ED25519:
		var pub = k.privateKey.(ed25519.PrivateKey).Public().(ed25519.PublicKey)
		pubKey, err = ssh.NewPublicKey(pub)

	default:
		err = ErrUnsupportedKeyType

	}

	return ssh.MarshalAuthorizedKey(pubKey), err

}


slowseeder/slowseeder.go
// Package slowseeder implements a drop-in replacement for a rand source
// intended for cryptographic key generation.
//
// It has been designed to be simple and reproducible. Generation is
// deterministic from a seed, uses multiple layered hashing functions,
// and is parameterized to easily extend the time spent during each
// iteration, making brute force and pre-computation more difficult.
//
package slowseeder

import (
	"crypto/sha512"
	"errors"
	"io"
	"sync"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/pbkdf2"
	"golang.org/x/crypto/ripemd160"
)

// Reader represents a drop in replacement for a rand source
type Reader struct {
	seed, salt, key      []byte
	rounds, time, memory uint32
	threads              uint8

	mu    *sync.RWMutex
	reads int
}

// New returns a Reader generator suitable for use with cryptographic functions
func New(seed []byte, rounds, time, memory uint32, threads uint8) (io.Reader, error) {

	var err error

	if len(seed) == 0 {
		err = errors.New("Reader seed not set")
	}

	if rounds < 1 {
		err = errors.New("Reader seeder requires rounds > 0")
	}

	if time < 1 {
		err = errors.New("Reader seeder requires time > 0")
	}

	if memory < 1 {
		err = errors.New("Reader seeder requires memory > 0")
	}

	if threads < 1 {
		err = errors.New("Reader seeder requires threads > 0")
	}

	return &Reader{
		seed:    seed,
		rounds:  rounds,
		time:    time,
		memory:  memory,
		threads: threads,
		mu:      &sync.RWMutex{},
	}, err

}

// Read implements a Reader that uses SHA512 and RIPEMD160 PBKDF2 to
// iteratively hash the seed and salt, which are supplied to Argon2 to
// generate the requested "entropy"
func (r *Reader) Read(p []byte) (int, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.seed = pbkdf2.Key(r.seed, r.key, int(r.rounds), sha512.Size, sha512.New)
	r.salt = pbkdf2.Key(r.salt, r.key, int(r.reads), ripemd160.Size, ripemd160.New)
	r.key = argon2.Key(r.seed, r.salt, r.time, r.memory, r.threads, uint32(len(p)))
	return copy(p, r.key), nil
}


deterministic/deterministic.go
package deterministic

import (
	"crypto/sha512"
	"errors"
	"sync"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/pbkdf2"
	"golang.org/x/crypto/ripemd160"
)

// Deterministic represents a drop in replacement for a rand source
type Deterministic struct {
	seed    []byte
	salt    []byte
	rounds  uint32
	time    uint32
	memory  uint32
	threads uint8
	reads   int
	mu      *sync.RWMutex
}

// New returns a Deterministic generator suitable for use with cryptographic functions
func New(seed, salt []byte, rounds, time, memory uint32, threads uint8) (*Deterministic, error) {

	var d = &Deterministic{
		seed:    seed,
		salt:    salt,
		rounds:  rounds,
		time:    time,
		memory:  memory,
		threads: threads,
		mu:      &sync.RWMutex{},
	}

	return d, d.verifyState()

}

func (d *Deterministic) verifyState() (err error) {

	if len(d.seed) == 0 {
		err = errors.New("Deterministic seed not set")
	}

	if len(d.salt) == 0 {
		err = errors.New("Deterministic salt not set")
	}

	if d.rounds < 1 {
		err = errors.New("Deterministic seeder requires rounds > 0")
	}

	if d.time < 1 {
		err = errors.New("Deterministic seeder requires time > 0")
	}

	if d.memory < 1 {
		err = errors.New("Deterministic seeder requires memory > 0")
	}

	if d.threads < 1 {
		err = errors.New("Deterministic seeder requires threads > 0")
	}

	return

}

// Read implements a Reader that uses SHA512 and RIPEMD160 PBKDF2 to
// iteratively hash the seed and salt, which are supplied to Argon2ti to
// generate the requested "entropy"
func (d *Deterministic) Read(p []byte) (int, error) {

	if err := d.verifyState(); err != nil {
		panic(err)
	}

	d.mu.Lock()
	defer d.mu.Unlock()

	d.reads++
	d.seed = pbkdf2.Key(d.seed, d.salt, int(d.rounds), sha512.Size, sha512.New)
	d.salt = pbkdf2.Key(d.salt, d.seed, int(d.rounds), ripemd160.Size, ripemd160.New)

	return copy(p, argon2.IDKey(d.seed, d.salt, d.time, d.memory, d.threads, uint32(len(p)))), nil
}

// Reads returns the number of reads that have occurred
func (d *Deterministic) Reads() int {
	d.mu.RLock()
	defer d.mu.RUnlock()
	return d.reads
}


main.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strings"

	"github.com/cornfeedhobo/ssh-keydgen/keygen"
	"github.com/cornfeedhobo/ssh-keydgen/slowseeder"
	"github.com/mitchellh/go-homedir"
	"golang.org/x/crypto/ed25519"
	"golang.org/x/crypto/ssh/agent"
	"golang.org/x/crypto/ssh/terminal"
	"gopkg.in/urfave/cli.v1"
)

func newError(message string) error {
	return cli.NewExitError(message, 1)
}

func newBug(message string) error {
	return cli.NewExitError(message, 13)
}

func main() {
	app := cli.NewApp()

	app.Name = "ssh-keygen"
	app.Version = "0.4.0"

	app.Author = "cornfeedhobo"
	app.Copyright = "(c) 2018 cornfeedhobo"

	app.HelpName = "ssh-keygen"
	app.Usage = "deterministic authentication key generation"
	app.UsageText = "ssh-keygen [[-t <type>] [-b <bits>] [-c <curve>] [-f <filename>] [-a <rounds>] [--at <time>] [--am <memory>] [--as <seedphrase>] [--aa]]"

	app.HideHelp = true
	app.HideVersion = true

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  "t",
			Value: "rsa",
			Usage: "Specifies the `type` of key to create. The possible values are \"dsa\", \"ecdsa\", \"rsa\", or \"ed25519\".",
		},
		cli.IntFlag{
			Name:  "b",
			Value: 2048,
			Usage: "Specifies the number of `bits` in the key to create. Possible values are restricted by key type.",
		},
		cli.IntFlag{
			Name:  "c",
			Value: 256,
			Usage: "Specifies the elliptic `curve` to use. The possible values are 256, 384, or 521.",
		},
		cli.StringFlag{
			Name:  "f",
			Usage: "Specifies the `filename` of the key file.",
		},
		cli.IntFlag{
			Name:  "a",
			Value: 1000,
			Usage: "Specifies the number of hashing `rounds` applied during key generation.",
		},
		cli.UintFlag{
			Name:  "at",
			Value: 3,
			Usage: "Specifies the `time` parameter for the Argon2 function.",
		},
		cli.UintFlag{
			Name:  "am",
			Value: 1024 * 16,
			Usage: "Specifies the `memory` parameter for the Argon2 function.",
		},
		cli.UintFlag{
			Name:  "ap",
			Value: 1,
			Usage: "Specifies the `threads` or parallelism for the Argon2 function.",
		},
		cli.StringFlag{
			Name:  "as",
			Usage: "Provides the deterministic `seedphrase`.",
		},
		cli.BoolFlag{
			Name:  "aa",
			Usage: "Add the generated key to the running ssh-agent.",
		},
	}

	app.Action = appAction

	app.Run(os.Args)

}

func appAction(ctx *cli.Context) (err error) {

	ctx.Set("t", strings.ToLower(ctx.String("t")))

	if ctx.Bool("aa") && os.Getenv("SSH_AUTH_SOCK") == "" {
		return newError("SSH_AUTH_SOCK not set, unable to find running agent")
	}

	fmt.Println("Generating public/private " + ctx.String("t") + " key pair")

	var seedphrase []byte
	if seedphrase, err = getSeedphrase(ctx); err != nil {
		return
	}

	var filename string
	if filename, err = getFilename(ctx); err != nil {
		return
	}

	var keydgen = &keygen.Keydgen{
		Type:  ctx.String("t"),
		Bits:  uint16(ctx.Int("b")),
		Curve: uint16(ctx.Int("c")),
	}

	rand, err := slowseeder.New(seedphrase, uint32(ctx.Int("a")), uint32(ctx.Uint("at")), uint32(ctx.Uint("am")), uint8(ctx.Uint("ap")))
	if err != nil {
		return newError("Error with supplied parameters: " + err.Error())
	}

	privateKey, err := keydgen.GenerateKey(rand)
	if err != nil {
		return newError("Error generating key: " + err.Error())
	}

	if ctx.Bool("aa") {
		err = addKeyToAgent(privateKey)
	} else {
		err = writeKeyToFile(keydgen, filename)
	}

	return

}

func getFilename(ctx *cli.Context) (filename string, err error) {

	filename = ctx.String("f")
	if !ctx.Bool("aa") && filename == "" {
		var home string
		home, err = homedir.Dir()
		if err != nil {
			err = newBug(err.Error())
			return
		}

		fmt.Printf("Enter file in which to save the key (%s/.ssh/id_%s): ", home, ctx.String("t"))
		if _, err = fmt.Scanln(&filename); err != nil {
			return
		}
	}

	abspath, err := filepath.Abs(filename)
	if err != nil {
		err = newError(err.Error())
		return
	}

	_, privStatErr := os.Stat(abspath)
	_, pubStatErr := os.Stat(abspath + ".pub")
	if privStatErr == nil || pubStatErr == nil {

		var strbool string
		fmt.Println(filename + " already exists.")
		fmt.Print("Overwrite (y/n)? ")
		if _, err = fmt.Scanln(&strbool); err != nil {
			return
		}

		if strings.TrimSpace(strings.ToLower(strbool)) == "y" {

			err = os.Remove(abspath)
			if err != nil && !os.IsNotExist(err) {
				return
			}

			err = os.Remove(abspath + ".pub")
			if err != nil && !os.IsNotExist(err) {
				return
			}

		} else {
			err = newError("")
		}
	}

	return

}

func getSeedphrase(ctx *cli.Context) (seed []byte, err error) {

	stat, _ := os.Stdin.Stat()

	if (stat.Mode() & os.ModeCharDevice) == 0 {

		seed, err = ioutil.ReadAll(os.Stdin)

	} else {

		var (
			equal bool
			fd    = int(os.Stdin.Fd())
		)

		for !equal {

			seed = []byte(ctx.String("as"))
			for len(seed) == 0 {
				fmt.Print("Enter seedphrase (can not be empty): ")
				seed, err = terminal.ReadPassword(fd)
				fmt.Print("\n")
				if err != nil {
					break
				}
			}

			verify := []byte(ctx.String("as"))
			for len(verify) == 0 {
				fmt.Print("Verify seedphrase (can not be empty): ")
				verify, err = terminal.ReadPassword(fd)
				fmt.Print("\n")
				if err != nil {
					break
				}
			}

			equal = bytes.Equal(seed, verify)
			if !equal {
				fmt.Print("\nerror: seedphrases did not match\n\n")
			}

		}

	}

	return

}

func addKeyToAgent(privateKey interface{}) error {

	conn, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK"))
	if err != nil {
		return err
	}

	// because agent.Client.Add() requires a pointer for all types
	if k, ok := privateKey.(ed25519.PrivateKey); ok {
		privateKey = &k
	}

	return agent.NewClient(conn).Add(agent.AddedKey{PrivateKey: privateKey})

}

func writeKeyToFile(k *keygen.Keydgen, filename string) error {

	privBytes, err := k.MarshalPrivateKey()
	if err != nil {
		return newError(err.Error())
	}

	pubBytes, err := k.MarshalPublicKey()
	if err != nil {
		return newError(err.Error())
	}

	err = ioutil.WriteFile(filename, privBytes, 0600)
	if err != nil {
		return newError(err.Error())
	}

	err = ioutil.WriteFile(filename+".pub", pubBytes, 0600)
	if err != nil {
		return newError(err.Error())
	}

	return nil

}
